### PR TITLE
+str #19041 deterministic `interleave` operation

### PR DESCRIPTION
**Pull request by agolubev**
Ref #19041

### CODE REVIEW

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowMergeSpec.scala [hunk 1]
```diff
@@ -26,7 +26,7 @@ class FlowMergeSpec extends BaseTwoStreamsSetup {
       val source3 = Source(List[Int]())
       val probe = TestSubscriber.manualProbe[Int]()
 
-      Source(0 to 3).merge(Source(List[Int]())).merge(Source(4 to 9))
+      source1.merge(source3).merge(source2)
```
Comment (viktorklang) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowMergeSpec.scala]: Isn't that a weird order?

File: akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala [hunk 2]
```diff
@@ -934,6 +934,48 @@ final class Flow[-In, +Out, +Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends
...
+   *
+   * '''Emits when''' element is available from current stream or from the given [[Source]] depending on what was pulled
+   *
+   * '''Backpressures when''' downstream backpressures
+   *
+   * '''Completes when''' the [[Flow]] and given [[Source]] completes
+   *
+   * '''Cancels when''' downstream cancels
+   */
+  def interleave[T >: Out](that: Graph[SourceShape[T], _], result: Int): javadsl.Flow[In, T, Mat] =
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: I think we need alternatives for `result` (the name). Perhaps `consecutive` or `sliceSize` or something like that.

File: akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala [hunk 3]
```diff
@@ -934,6 +934,48 @@ final class Flow[-In, +Out, +Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends
     new Flow(delegate.alsoToMat(that)(combinerToScala(matF)))
 
   /**
+   * Interleave represents deterministic merge of the given [[Source]] with elements of this [[Flow]].
+   * It takes `request` number of elements from this flow to emit downstream, then - same amount for given source,
+   * then repeat process from the beginning.
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: Perhaps:

It first emits `request` number of elements from this flow to downstream, then - same amount for `that` source, then repeat process.

File: akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala [hunk 4]
```diff
@@ -934,6 +934,48 @@ final class Flow[-In, +Out, +Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends
     new Flow(delegate.alsoToMat(that)(combinerToScala(matF)))
 
   /**
+   * Interleave represents deterministic merge of the given [[Source]] with elements of this [[Flow]].
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: `Interleave is a deterministic merge `

File: akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala [hunk 5]
```diff
@@ -934,6 +934,48 @@ final class Flow[-In, +Out, +Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends
...
+   * Source<Integer, ?> src = Source.from(Arrays.asList(1, 2, 3))
+   * Flow<Integer, Integer, ?> flow = flow.interleave(Source.from(Arrays.asList(4, 5, 6, 7)), 2)
+   * src.via(flow) // 1, 2, 4, 5, 3, 6, 7
+   * }}}
+   *
+   * After [[Flow]] or [[Source]] is complete than all the rest elements will be emitted from the second one
+   *
+   * If this [[Flow]] or [[Source]] gets upstream error - stream completes with failure.
+   *
+   * '''Emits when''' element is available from current stream or from the given [[Source]] depending on what was pulled
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: I find `current` confusing when contrasted with `given` here.
What do you think about:

"element is available from the currently consumed upstream"

File: akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala [hunk 6]
```diff
@@ -934,6 +934,48 @@ final class Flow[-In, +Out, +Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends
...
+   *
+   * '''Backpressures when''' downstream backpressures
+   *
+   * '''Completes when''' the [[Flow]] and given [[Source]] completes
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: Since this is on FlowOps, [[Flow]] and [[Source]] becomes ambiguous or potentially misleading. I'd suggest consistently referring them to "`this`'s upstream or `that`".

File: akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala [hunk 7]
```diff
@@ -225,6 +225,87 @@ class MergePreferred[T] private (val secondaryPorts: Int, val eagerClose: Boolea
...
+ * '''Emits when''' element is available from current input (depending on phase)
+ *
+ * '''Backpressures when''' downstream backpressures
+ *
+ * '''Completes when''' all upstreams complete (eagerClose=false) or one upstream completes (eagerClose=true)
+ *
+ * '''Cancels when''' downstream cancels
+ *
+ */
+class Interleave[T] private (val inputPorts: Int, val request: Int, val eagerClose: Boolean) extends GraphStage[UniformFanInShape[T, T]] {
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: `final class`

**Comment by ktoso** [comment]
OK TO TEST

File: akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala [hunk 8]
```diff
@@ -934,6 +934,48 @@ final class Flow[-In, +Out, +Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends
...
+   * src.via(flow) // 1, 2, 4, 5, 3, 6, 7
+   * }}}
+   *
+   * After [[Flow]] or [[Source]] is complete than all the rest elements will be emitted from the second one
+   *
+   * If this [[Flow]] or [[Source]] gets upstream error - stream completes with failure.
+   *
+   * '''Emits when''' element is available from current stream or from the given [[Source]] depending on what was pulled
+   *
+   * '''Backpressures when''' downstream backpressures
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: Was recently thinking about that one in most docs; It does not sound very useful for fanout / fanin ops, 
since it does not explain if "any of the two backpressures" or if "both backpressure", can we reword this to explain the semantics we have here?

File: akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala (hunk 7 above)
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: This has to make sure that `request > 0` otherwise the logic will get messed up.

Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: Also, require `inputPorts > 1` (see Merge and friends for example)

File: akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala [hunk 9]
```diff
@@ -232,9 +313,8 @@ object Broadcast {
    * @param outputPorts number of output ports
    * @param eagerCancel if true, broadcast cancels upstream if any of its downstreams cancel.
    */
-  def apply[T](outputPorts: Int, eagerCancel: Boolean = false): Broadcast[T] = {
+  def apply[T](outputPorts: Int, eagerCancel: Boolean = false): Broadcast[T] =
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: Thanks! :)

File: akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala (hunk 7 above)
Comment (hepin1989) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: require for safety:)

File: akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala (hunk 2 above)
Comment (agolubev) [inline akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: @viktorklang, for RX they called it count (in many places). is it good enough?

File: akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala (hunk 5 above)
Comment (agolubev) [inline akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: yeah, sounds natural

File: akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala (hunk 7 above)
Comment (agolubev) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: @viktorklang, none of fan in/out classes in Graph.scala is final (Merge, Broadcast) - should I make them all final?

Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: Looks very final to me: https://github.com/akka/akka/blob/release-2.3-dev/akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala#L38

File: akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala (hunk 2 above)
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: Count is good but a bit ambiguous since it could refer to a total count rather than a segment size.

File: akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala (hunk 7 above)
Comment (agolubev) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: yeah, it's changed since I brunched out. Ok.

File: akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala (hunk 2 above)
Comment (agolubev) [inline akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: segmentSize than?

Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: `segmentSize` sounds good to me!

**Comment by johanandren** [comment]
PLS BUILD

**Comment by agolubev** [comment]
will rebase

**Comment by agolubev** [comment]
Ok, Ready for review

**Comment by johanandren** [comment]
PLS BUILD

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowInterleaveSpec.scala [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+    val subscriber = TestSubscriber.probe[Outputs]()
+    Source(p1).interleave(Source(p2), 2).runWith(Sink(subscriber))
+    subscriber
+  }
+
+  "An Interleave for Flow " must {
+
+    "work in the happy case" in assertAllStagesStopped {
+      val probe = TestSubscriber.manualProbe[Int]()
+      Source(0 to 3).interleave(Source(List[Int]()), 2).interleave(Source(4 to 9), 2).runWith(Sink(probe))
```
Comment (viktorklang) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowInterleaveSpec.scala]: Would be nice with a test where the `segmentSize` is different for each of the `interleaves`

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowInterleaveSpec.scala [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+    "work when bucket is not equal elements in stream" in assertAllStagesStopped {
+      val probe = TestSubscriber.manualProbe[Int]()
+
+      Source(0 to 2).interleave(Source(3 to 5), 2).runWith(Sink(probe))
+      probe.expectSubscription().request(10)
+      probe.expectNext(0, 1, 3, 4, 2, 5)
+      probe.expectComplete()
+    }
+
+    "work with bucket = 1" in assertAllStagesStopped {
```
Comment (viktorklang) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowInterleaveSpec.scala]: segmentSize?

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowInterleaveSpec.scala [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+      for (_ ← 1 to 10) {
+        subscription.request(1)
+        collected += probe.expectNext()
+      }
+
+      collected should be(Set(0, 1, 4, 5, 2, 3, 6, 7, 8, 9))
+      probe.expectComplete()
+    }
+
+    "work when bucket is not equal elements in stream" in assertAllStagesStopped {
```
Comment (viktorklang) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowInterleaveSpec.scala]: segmentSize?

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowInterleaveSpec.scala [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+
+      Source(0 to 2).interleave(Source(3 to 5), 1).runWith(Sink(probe))
+      probe.expectSubscription().request(10)
+      probe.expectNext(0, 3, 1, 4, 2, 5)
+      probe.expectComplete()
+    }
+
+    commonTests()
+
+    "work with one immediately completed and one nonempty publisher" in assertAllStagesStopped {
```
Comment (viktorklang) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowInterleaveSpec.scala]: A test that exersizes that exceptions are thrown when illegal segmentSize is used?

Comment (viktorklang) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowInterleaveSpec.scala]: Also, a test that exersizes the case where the segmentSize is larger than one or both of the upstream sizes?

**Comment by agolubev** [comment]
Added, all that was requested in feedback

File: akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala [hunk 14]
```diff
@@ -227,6 +227,90 @@ final class MergePreferred[T] private (val secondaryPorts: Int, val eagerClose:
...
+                if (isAvailable(out)) pull(currentUpstream)
+              }
+            } else completeStage()
+          }
+        }
+      })
+    }
+
+    setHandler(out, new OutHandler {
+      override def onPull(): Unit = if (!hasBeenPulled(currentUpstream)) pull(currentUpstream)
```
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: please use tryPull here, the port could already be closed but the notification not yet delivered

File: akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala [hunk 15]
```diff
@@ -227,6 +227,90 @@ final class MergePreferred[T] private (val secondaryPorts: Int, val eagerClose:
...
+      setHandler(i, new InHandler {
+        override def onPush(): Unit = {
+          push(out, grab(i))
+          counter += 1
+          if (counter == segmentSize) switchToNextInput()
+        }
+
+        override def onUpstreamFinish() = {
+          if (eagerClose) {
+            in.foreach(cancel)
```
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: this is done automatically by completeStage()

File: akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala [hunk 16]
```diff
@@ -227,6 +227,90 @@ final class MergePreferred[T] private (val secondaryPorts: Int, val eagerClose:
...
+    private var counter = 0
+    private var currentUpstreamIndex = 0
+    private var runningUpstreams = inputPorts
+    private def upstreamsClosed = runningUpstreams == 0
+    private def currentUpstream = in(currentUpstreamIndex)
+
+    private def switchToNextInput(): Unit = {
+      @tailrec
+      def nextInletIndex(index: Int): Int = {
+        val reduced = (index + 1) % inputPorts
```
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: `%` is expensive, I’d rather say
~~~scala
(index + 1) match {
  case `inputPorts` => 0
  case x => x
}
~~~

File: akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala [hunk 17]
```diff
@@ -227,6 +227,90 @@ final class MergePreferred[T] private (val secondaryPorts: Int, val eagerClose:
...
+        if (!isClosed(in(reduced))) reduced else nextInletIndex(index + 1)
```
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: `nextInletIndex(reduced)`

Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: … and I’d probably name it “successor” instead of “reduced” ;-)

Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: Another thing: `isClosed` may return true for a port for which `onUpstreamFinish` has not yet been executed, meaning that this needs a termination condition of `if (successor == currentUpstreamIndex) completeStage()`

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowInterleaveSpec.scala [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+
+  "An Interleave for Flow " must {
+
+    "work in the happy case" in assertAllStagesStopped {
+      val probe = TestSubscriber.manualProbe[Int]()
+      Source(0 to 3).interleave(Source(List[Int]()), 2).interleave(Source(4 to 9), 3).runWith(Sink(probe))
+
+      val subscription = probe.expectSubscription()
+
+      var collected = Set.empty[Int]
```
Comment (rkuhn) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowInterleaveSpec.scala]: since this merge shall be deterministic we shouldn’t need Sets, but Seq should work, right?

**Comment by rkuhn** [comment]
Great work, @agolubev !

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowInterleaveSpec.scala (hunk 18 above)
Comment (agolubev) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowInterleaveSpec.scala]: yes, missed that.

File: akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala (hunk 16 above)
Comment (agolubev) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: hm - looks interesting.

File: akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala (hunk 17 above)
Comment (agolubev) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: ouch - good to know.

Comment (agolubev) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: @rkuhn,  ```switchToNextInput``` is called only from ```onPush``` when it's possible to have in port closed without ```onUpstreamFinish```. (another call is from ```onUpstreamFinish``` itself but in this case mentioned state iss not possible I believe). My concern is that we are calling ```push(out, grab(i))``` and then ```completeStage()```. Will it discard previous push?

Comment (agolubev) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: found ```push()``` and then ```completeStage``` in ```GroupedWithin.emitGroup```.

**Comment by rkuhn** [comment]
fix for the test failure is in #19164
